services/skus/appstore.go

```diff
@@ -0,0 +1,308 @@
+	case x.val.NotificationType == appstore.NotificationTypeV2Expired && x.val.Subtype == appstore.SubTypeV2BillingRetry:
+		return true
+
+	// Cancellation after price increase.
+	case x.val.NotificationType == appstore.NotificationTypeV2DidChangeRenewalStatus && x.val.Subtype == "":
+		return true
+
+	default:
+		return false
+	}
+}
+
+func (x *appStoreSrvNotification) effect() string {
+	switch {
+	case x.shouldRenew():
+		return "renew"
+	case x.shouldCancel():
+		return "cancel"
+	default:
+		return "skip"
+	}
+}
+
+func parseAppStoreSrvNotification(vrf *assnCertVerifier, spayload string) (*appStoreSrvNotification, error) {
+	certs, err := extractASSNCerts(spayload)
+	if err != nil {
+		return nil, err
+	}
+
+	if len(certs) != 3 {
+		return nil, errInvalidASSNCertNum
+	}
+
+	if err := vrf.verify(certs[2], certs[1]); err != nil {
+		return nil, err
+	}
+
+	pubKey, err := extractPubKey(certs[0])
+	if err != nil {
+		return nil, err
+	}
+
+	raw, err := jose.ParseSigned(spayload)
+	if err != nil {
+		return nil, err
+	}
+
+	payload, err := raw.Verify(pubKey)
+	if err != nil {
+		return nil, err
+	}
+
+	ntf := &appstore.SubscriptionNotificationV2DecodedPayload{}
+	if err := json.Unmarshal(payload, ntf); err != nil {
+		return nil, err
+	}
+
+	result := &appStoreSrvNotification{
+		pubKey: pubKey,
+		val:    ntf,
+	}
+
+	return result, nil
+}
+
+func parseTxnInfo(pubKey *ecdsa.PublicKey, spayload appstore.JWSTransaction) (*appstore.JWSTransactionDecodedPayload, error) {
+	raw, err := jose.ParseSigned(string(spayload))
+	if err != nil {
+		return nil, err
+	}
+
+	data, err := raw.Verify(pubKey)
+	if err != nil {
+		return nil, err
+	}
+
+	result := &appstore.JWSTransactionDecodedPayload{}
+	if err := json.Unmarshal(data, result); err != nil {
+		return nil, err
+	}
+
+	return result, nil
+}
+
+//nolint:unused
+func parseRenewalInfo(pubKey *ecdsa.PublicKey, spayload appstore.JWSRenewalInfo) (*appstore.JWSRenewalInfoDecodedPayload, error) {
+	raw, err := jose.ParseSigned(string(spayload))
+	if err != nil {
+		return nil, err
+	}
+
+	data, err := raw.Verify(pubKey)
+	if err != nil {
+		return nil, err
+	}
+
+	result := &appstore.JWSRenewalInfoDecodedPayload{}
+	if err := json.Unmarshal(data, result); err != nil {
+		return nil, err
+	}
+
+	return result, nil
+}
+
+type assnCertVerifier struct {
+	root *x509.CertPool
+}
+
+func newASSNCertVerifier() (*assnCertVerifier, error) {
+	pool := x509.NewCertPool()
+	if ok := pool.AppendCertsFromPEM([]byte(appleRootCert)); !ok {
+		return nil, errFailedParseASCert
+	}
+
+	result := &assnCertVerifier{root: pool}
+
+	return result, nil
+}
+
+func (x *assnCertVerifier) verify(rcert, icert *x509.Certificate) error {
+	interm := x509.NewCertPool()
+	interm.AddCert(icert)
+
+	opts := x509.VerifyOptions{
+		Roots:         x.root,
+		Intermediates: interm,
+	}
+
+	if _, err := rcert.Verify(opts); err != nil {
+		return err
+	}
+
+	// Consider also verifying the public key.
```

Comment:
See here for old implementation that matches: https://github.com/brave-intl/bat-go/pull/2483/files#diff-6ebc9187728cc8ddacabb1ef3b11220d2fab15950be2e72861dd9e93ee171268L270-L300

Reply:
The old implementation did not check the public key. They only checked the root cert (the third item in the array) against the Apple Root Cert and the intermediary.

The new equivalent code is here https://github.com/brave-intl/bat-go/pull/2483/files#diff-442296cd497870722e77d2e12675b7734d2bd9c7cc57106969d16020ef57e27dR237-R248.

Reply:
Yeah makes sense, that's why I didn't want to block this was that we weren't changing anything. I'm going to look into the docs though and see how they handle key management for this to see if we do need to verify the public key in any sort of way or if this is fine. I presume because this was how it was implemented previously that this is fine, but it's not common to leave the public key unverified typically. However, this is still being signed by the root CA so we know for certain the key is coming from Apple and/or attested to by them. The rest is to figure out what the meaning of this key is and how they're generating it.

Reply:
My plan was to ship this, and then add the check. Simply to avoid blowing up the possible scope of errors.

Reply:
Adding that check is as simple as just passing it to `verify` along with the root and interm certs. Basically, all three should be passed, and then after the root, I'll just also verify the leaf cert.

Pretty much like it's done here – https://github.com/awa/go-iap/blob/14fdbb2d2b3873e9939e56b741f26e233034334d/appstore/cert.go#L84-L87.